src/wiki/forms_account_handling.py

```diff
@@ -0,0 +1,89 @@
+import random
+import string
+
+import django.contrib.auth.models
+from django import forms
+from django.contrib.auth import get_user_model
+from django.contrib.auth.forms import UserCreationForm
+from django.core.exceptions import FieldDoesNotExist
+from django.db.models.fields import CharField, EmailField
+from django.utils.translation import gettext_lazy as _
+from wiki.conf import settings
+
+
+def _get_field(model, field):
+    try:
+        return model._meta.get_field(field)
+    except FieldDoesNotExist:
+        return
+
+
+User = get_user_model()
+
+
+def check_user_field(user_model=User):
+    return isinstance(_get_field(user_model, user_model.USERNAME_FIELD), CharField)
+
+
+def check_email_field(user_model=User):
+    return isinstance(_get_field(user_model, user_model.get_email_field_name()), EmailField)
+
+
+# django parses the ModelForm (and Meta classes) on class creation, which fails with custom models without expected fields.
+# We need to check this here, because if this module can't load then system checks can't run.
+CustomUser = User \
+    if (settings.ACCOUNT_HANDLING and check_user_field() and check_email_field()) \
+    else django.contrib.auth.models.User
```

Comment:
Is it a good idea that this silently fails if some checks don't match up? If `user_model.USERNAME_FIELD` is set to something that isn't valid, then expected behavior wouldn't be to just silently assume a different model?

Reply:
I'll take a closer look. I agree, I'm suspecting I didn't want to break any existing behaviour.

Reply:
Getting my head back around this again. Hopefully it doesn't fail silently. `tests.core.test_checks.CheckTests.test_check_for_fields_in_custom_user_model` should alert the user if there is a problem.
The reason for the `CustomUser` test is that Django/six do some pre-processing on `UserCreationForm/UserUpdateForm` at the time the class is **defined**, not used, so if there is a problem with the User model, forms_account_handling doesn't load, the above check doesn't run, and the user gets an exception from the django library, rather than a more helpful one from django-wiki checks.

Note also that `tests.core.test_checks.CheckTests.test_custom_user_model_mitigation_required` tests for the requirement for the `CustomUser` test, to alert us if django changes its behavior in the future.
This documentation accompanies the check:
```
Django & six check django.forms.ModelForm.Meta on definition, and raises an error if Meta.fields don't exist in Meta.model.
This causes problems in wiki.forms_account_handling.UserCreationForm and wiki.forms_account_handling.UserUpdateForm when a custom user model doesn't have fields django-wiki assumes.
There is some code in wiki.forms that detects this situation.
This check asserts that Django/six are still raising an exception on definition, and asserts the mitigation code in wiki.forms, and that test_check_for_fields_in_custom_user_model below are required.
```

Without this test, this exception is raised:
```
Traceback (most recent call last):
  File "./src-custom-model/manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/lib/python3.5/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/lib/python3.5/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/lib/python3.5/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/lib/python3.5/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/lib/python3.5/site-packages/django/core/management/commands/check.py", line 68, in handle
    fail_level=getattr(checks, options['fail_level']),
  File "/lib/python3.5/site-packages/django/core/management/base.py", line 359, in check
    include_deployment_checks=include_deployment_checks,
  File "/lib/python3.5/site-packages/django/core/management/base.py", line 346, in _run_checks
    return checks.run_checks(**kwargs)
  File "/lib/python3.5/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/lib/python3.5/site-packages/django/core/checks/urls.py", line 16, in check_url_config
    return check_resolver(resolver)
  File "/lib/python3.5/site-packages/django/core/checks/urls.py", line 26, in check_resolver
    return check_method()
  File "/lib/python3.5/site-packages/django/urls/resolvers.py", line 256, in check
    for pattern in self.url_patterns:
  File "/lib/python3.5/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/lib/python3.5/site-packages/django/urls/resolvers.py", line 407, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/lib/python3.5/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/lib/python3.5/site-packages/django/urls/resolvers.py", line 400, in urlconf_module
    return import_module(self.urlconf_name)
  File "/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "./src-custom-model/testproject/urls.py", line 26, in <module>
    url(r'', include('wiki.urls'))
  File "/lib/python3.5/site-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "./django-wiki/src/wiki/urls.py", line 6, in <module>
    from wiki.views import accounts, article, deleted_list
  File "./django-wiki/src/wiki/views/accounts.py", line 21, in <module>
    from wiki import forms
  File "./django-wiki/src/wiki/forms.py", line 39, in <module>
    from .forms_account_handling import UserCreationForm, UserUpdateForm
  File "./django-wiki/src/wiki/forms_account_handling.py", line 71, in <module>
    class UserUpdateForm(forms.ModelForm):
  File "/lib/python3.5/site-packages/django/forms/models.py", line 277, in __new__
    raise FieldError(message)
django.core.exceptions.FieldError: Unknown field(s) (email) specified for MyUser
```

With the test we can avoid that and produce our own
```
SystemCheckError: System check identified some issues:

ERRORS:
testapp.MyUser: (wiki.E011) testapp.models.MyUser.get_email_field_name() refers to a field that is not of type EmailField
	HINT: If you have your own login/logout views, turn off settings.WIKI_ACCOUNT_HANDLING

System check identified 1 issue (0 silenced).
```

The situation is far from ideal, containing it in forms_account_handling with that check is the best I've been able to come up with. @benjaoming please let me know what you think?

Reply:
I had to read the explanation two times, but it was well-worth it. Thanks for explaining it. It's actually an interesting case: Having to adapt "models.py" to avoid normal exceptional states in order to raise the exceptional state from "checks.py". It seems like an unfortunate side-effect of the checks framework.

Thanks a lot, I'm good with the chosen solution. The advantage is as you say that the checks message is more helpful.